File: ts/src/token-manager.ts
/**
 * Token manager for long-lived, single-identity clients — SPEC §4.2.1.
 *
 * A convenience wrapper over `auth.token()` (§4.2) for desktop apps, sync
 * agents, and daemons that hold one refresh token and need a
 * continuously-valid access token without hand-rolling the refresh loop.
 * It is NOT for browser/BFF flows (use the middleware, §10) nor for the
 * SDK's internal platform session (§4.0).
 *
 * Rotation safety: user refresh tokens are one-time-use and rotate on
 * every /auth/token (ADR-031). The manager therefore:
 *   (a) single-flights concurrent acquisitions so two callers never present
 *       the same refresh token in parallel (which would trip
 *       reuse-detection and kill the session), and
 *   (b) persists the rotated token through the `refreshSink` *before*
 *       returning the new access token, so a crash can never strand the
 *       client on a consumed token.
 *
 * Construct via `realm.auth.newTokenManager(refreshToken, opts)`.
 */

import type { AuthClient } from "./auth.js";
import { RealmError } from "./errors.js";

/**
 * Durably persists a rotated refresh token. The manager calls it after each
 * successful /auth/token and AWAITS it: only if it resolves does the manager
 * cache and hand back the new access token. A sink that rejects fails the
 * acquisition (the caller's durable store still holds the previous token).
 * Best-effort / fire-and-forget sinks do NOT satisfy the crash-safety
 * contract (SPEC §4.2.1).
 */
export type RefreshSink = (newRefreshToken: string) => Promise<void> | void;

export interface TokenManagerOptions {
  /**
   * Tenant id sent on each /auth/token (required for multi-tenant user
   * refresh tokens; ignored otherwise).
   */
  tenantId?: string;
  /** Durable sink for rotated refresh tokens. See {@link RefreshSink}. */
  refreshSink?: RefreshSink;
  /** Clock override (tests only). Returns epoch milliseconds. */
  clock?: () => number;
}

/**
 * How far ahead of expiry the manager proactively refreshes, so callers
 * under steady load never observe an expired access token.
 */
const REFRESH_LEAD_MS = 60 * 1000;

/** Default access-token lifetime when the server omits expires_in (SPEC §4.0). */
const DEFAULT_ACCESS_TTL_MS = 5 * 60 * 1000;

interface InFlight {
  promise: Promise<string>;
}

export class TokenManager {
  private readonly auth: AuthClient;
  private readonly tenantId: string;
  private readonly sink?: RefreshSink;
  private readonly now: () => number;

  private refreshTokenValue: string;
  private cachedAccessToken = "";
  private accessExpiresAtMs = 0;
  private inflight: InFlight | null = null;
  /**
   * The access token the LAST forced refresh produced. The whole of D13's
   * bookkeeping: a `token_stale` on this exact token means the refresh already
   * happened and did not help, so refreshing again would be the loop. One
   * string, not a set — the cap is per token, and the only token that can be
   * "already refreshed for" is the one we just minted.
   */
  private forcedToken = "";

  constructor(auth: AuthClient, refreshToken: string, opts: TokenManagerOptions = {}) {
    this.auth = auth;
    this.refreshTokenValue = refreshToken;
    this.tenantId = opts.tenantId ?? "";
    this.sink = opts.refreshSink;
    this.now = opts.clock ?? (() => Date.now());
  }

  /**
   * Returns a valid access token, refreshing on demand. Returns the cached
   * token while it has at least 60s of life; otherwise performs (or joins)
   * a single refresh. A `refresh_invalid` error (SPEC §3.1) is terminal —
   * the caller should discard its stored refresh token and re-run its
   * login/enrollment flow.
   */
  async accessToken(): Promise<string> {
    if (this.cachedAccessToken !== "" && this.accessExpiresAtMs - this.now() >= REFRESH_LEAD_MS) {
      return this.cachedAccessToken;
    }
    // Join an in-flight acquisition rather than starting a second one — two
    // parallel /auth/token calls on the same one-time-use refresh token
    // would trip reuse-detection.
    if (this.inflight) {
      return this.inflight.promise;
    }
    const refresh = this.refreshTokenValue;
    const promise = this.fetch(refresh).finally(() => {
      this.inflight = null;
    });
    this.inflight = { promise };
    return promise;
  }

  /**
   * Answers an ADR-107 `token_stale` 401 by forcing ONE refresh and returning
   * the token to replay the original request with.
   *
   * D13, the loop-breaker. D8 makes the C5 loop very unlikely by stamping the
   * marker early; this makes it impossible, and it lives in the client SDK
   * because that is where the retry decision is actually made. A second
   * `token_stale` on a token this method already produced is a HARD failure —
   * rejected as `token_stale` so the caller can surface it, never retried.
   *
   * `staleToken` is the token the failing request carried. Passing one that is
   * no longer current is not an error: another caller already refreshed, and
   * you simply get the newer token to replay with, costing the issuer nothing.
   *
   * ```ts
   * let tok = await tm.accessToken();
   * try {
   *   return await call(tok);
   * } catch (e) {
   *   if (!isTokenStale(e)) throw e;
   *   tok = await tm.handleStale(tok);   // hard-rejects on the second try
   *   return await call(tok);            // surface a rejection; do NOT loop
   * }
   * ```
   */
  async handleStale(staleToken: string): Promise<string> {
    if (staleToken !== "" && staleToken === this.forcedToken) {
      throw new RealmError({
        code: "token_stale",
        httpStatus: 401,
        message:
          "realmid: token_stale on a token minted by the previous forced refresh — " +
          "refusing to refresh again (ADR-107 D13). The authority marker is ahead of " +
          "the issuer's clock, or the subject is being re-marked faster than a token " +
          "can be used.",
      });
    }
    // Someone else already refreshed past the failing token — replay with what
    // we hold rather than spending a mint.
    if (staleToken !== "" && this.cachedAccessToken !== "" && staleToken !== this.cachedAccessToken) {
      return this.cachedAccessToken;
    }
    // Drop the cached token so accessToken() cannot hand back the stale one.
    this.cachedAccessToken = "";
    this.accessExpiresAtMs = 0;
    const tok = await this.accessToken();
    this.forcedToken = tok;
    return tok;
  }

  /**
   * The manager's current refresh token (the latest rotated value). Useful
   * for a final persist on graceful shutdown.
   */
  refreshToken(): string {
    return this.refreshTokenValue;
  }

  /**
   * Performs one /auth/token acquisition. Only ever called by the
   * single-flight leader in {@link accessToken}.
   */
  private async fetch(refresh: string): Promise<string> {
    // refresh_invalid (and any other error) bubbles up unchanged: a
    // long-lived user client has no API key to fall back on, so a dead
    // refresh is terminal — re-authentication is required.
    const res = await this.auth.token({ refreshToken: refresh, tenantId: this.tenantId });

    if (!res.accessToken) {
      throw new RealmError({
        code: "server_error",
        message: "/auth/token returned empty access token",
      });
    }

    // User refresh always rotates; guard anyway so a non-rotating realm
    // config (service/platform) doesn't blank the stored token.
    const rotated = res.refreshToken || refresh;

    // Commit the rotated token to memory BEFORE persisting, so if the sink
    // fails and the caller retries, the retry presents the live (just
    // issued, unconsumed) token rather than the consumed one.
    this.refreshTokenValue = rotated;

    // Persist-before-return: the sink must durably store the rotated token
    // before we hand the caller an access token to use (SPEC §4.2.1). If it
    // rejects, fail the acquisition — do not cache the access token.
    if (this.sink) {
      try {
        await this.sink(rotated);
      } catch (e) {
        throw new RealmError({
          code: "server_error",
          message: "realmid token manager: persist rotated refresh token failed",
          cause: e,
        });
      }
    }

    const ttlMs = res.expiresIn > 0 ? res.expiresIn * 1000 : DEFAULT_ACCESS_TTL_MS;
    this.cachedAccessToken = res.accessToken;
    this.accessExpiresAtMs = this.now() + ttlMs;
    return res.accessToken;
  }
}

File: go/token_manager.go
package realmid

import (
	ctxpkg "context"
	"fmt"
	"sync"
	"time"
)

// TokenManager keeps a single user/identity session's access token fresh
// for a long-lived, single-identity client — a desktop app, sync agent,
// or daemon that holds one refresh token and needs a continuously-valid
// access token without hand-rolling the refresh loop (SPEC §4.2.1).
//
// It is NOT for browser/BFF flows (use the middleware, SPEC §10) nor for
// the SDK's internal platform session (§4.0). Construct it from a refresh
// token the client already holds via realm.Auth.NewTokenManager.
//
// Rotation safety: user refresh tokens are one-time-use and rotate on
// every /auth/token (ADR-031). The manager therefore (a) single-flights
// concurrent acquisitions so two goroutines never present the same
// refresh token (which would trip reuse-detection and kill the session),
// and (b) persists the rotated token through the RefreshSink *before*
// returning the new access token, so a crash can never strand the client
// on a consumed token.
type TokenManager struct {
	auth     *AuthClient
	tenantID string
	sink     RefreshSink
	now      func() time.Time

	mu              sync.Mutex
	refreshToken    string
	accessToken     string
	accessExpiresAt time.Time
	inflight        *tokenCall
	// forcedToken is the access token the LAST forced refresh produced. It is
	// the whole of D13's bookkeeping: a token_stale on this exact token means
	// the refresh already happened and did not help, so refreshing again would
	// be the loop. One string, not a set — the cap is per token, and the only
	// token that can be "already refreshed for" is the one we just minted.
	forcedToken string
}

// RefreshSink durably persists a rotated refresh token. The manager calls
// it after each successful /auth/token and BLOCKS on it: only if it
// returns nil does the manager cache and hand back the new access token.
// A sink that returns an error fails the acquisition (the caller's
// durable store still holds the previous token). Best-effort /
// fire-and-forget sinks do NOT satisfy the crash-safety contract.
type RefreshSink func(ctx ctxpkg.Context, newRefreshToken string) error

// TokenManagerOption configures a TokenManager at construction.
type TokenManagerOption func(*TokenManager)

// WithRefreshSink registers a durable sink for rotated refresh tokens.
// See RefreshSink for the persist-before-return contract.
func WithRefreshSink(sink RefreshSink) TokenManagerOption {
	return func(m *TokenManager) { m.sink = sink }
}

// WithTokenManagerTenant pins the tenant_id sent on each /auth/token
// (required for multi-tenant user refresh tokens; ignored otherwise).
func WithTokenManagerTenant(tenantID string) TokenManagerOption {
	return func(m *TokenManager) { m.tenantID = tenantID }
}

// withTokenManagerClock overrides the clock (tests only).
func withTokenManagerClock(now func() time.Time) TokenManagerOption {
	return func(m *TokenManager) {
		if now != nil {
			m.now = now
		}
	}
}

// tokenManagerRefreshLead is how far ahead of expiry the manager proactively
// refreshes, so callers under steady load never observe an expired token.
const tokenManagerRefreshLead = 60 * time.Second

// NewTokenManager builds a TokenManager seeded with a refresh token the
// client already holds (obtained out-of-band, e.g. at enrollment). The
// manager refreshes against POST /auth/token directly on that token.
func (a *AuthClient) NewTokenManager(refreshToken string, opts ...TokenManagerOption) *TokenManager {
	m := &TokenManager{
		auth:         a,
		refreshToken: refreshToken,
		now:          time.Now,
	}
	for _, opt := range opts {
		opt(m)
	}
	return m
}

// AccessToken returns a valid access token, refreshing on demand. It
// returns the cached token while it has at least tokenManagerRefreshLead
// of life; otherwise it performs (or joins) a single refresh. A
// refresh_invalid error (SPEC §3.1) is terminal — the caller should
// discard its stored refresh token and re-run its login/enrollment flow.
func (m *TokenManager) AccessToken(ctx ctxpkg.Context) (string, error) {
	m.mu.Lock()
	if m.accessToken != "" && m.accessExpiresAt.Sub(m.now()) >= tokenManagerRefreshLead {
		tok := m.accessToken
		m.mu.Unlock()
		return tok, nil
	}
	// Join an in-flight acquisition rather than starting a second one —
	// two parallel /auth/token calls on the same one-time-use refresh
	// would trip reuse-detection.
	if call := m.inflight; call != nil {
		m.mu.Unlock()
		<-call.done
		return call.token, call.err
	}
	call := &tokenCall{done: make(chan struct{})}
	m.inflight = call
	refresh := m.refreshToken
	m.mu.Unlock()

	tok, err := m.fetch(ctx, refresh)

	m.mu.Lock()
	m.inflight = nil
	m.mu.Unlock()
	call.token, call.err = tok, err
	close(call.done)
	return tok, err
}

// RefreshToken returns the manager's current refresh token (the latest
// rotated value). Useful for a final persist on graceful shutdown.
func (m *TokenManager) RefreshToken() string {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.refreshToken
}

// fetch performs one /auth/token acquisition. Only ever called by the
// single-flight leader in AccessToken.
func (m *TokenManager) fetch(ctx ctxpkg.Context, refresh string) (string, error) {
	mr, err := m.auth.Token(ctx, TokenRequest{RefreshToken: refresh, TenantID: m.tenantID})
	if err != nil {
		// refresh_invalid (and any other error) bubbles up unchanged: a
		// long-lived user client has no API key to fall back on, so a dead
		// refresh is terminal — re-authentication is required.
		return "", err
	}
	if mr.AccessToken == "" {
		return "", &RealmError{Code: ErrCodeServerError, Message: "/auth/token returned empty access token"}
	}
	rotated := mr.RefreshToken
	if rotated == "" {
		// User refresh always rotates; guard anyway so a non-rotating realm
		// config (service/platform) doesn't blank the stored token.
		rotated = refresh
	}
	// Commit the rotated token to memory BEFORE persisting, so if the sink
	// fails and the caller retries, the retry presents the live (just
	// issued, unconsumed) token rather than the consumed one.
	m.mu.Lock()
	m.refreshToken = rotated
	m.mu.Unlock()

	// Persist-before-return: the sink must durably store the rotated token
	// before we hand the caller an access token to use (SPEC §4.2.1).
	if m.sink != nil {
		if serr := m.sink(ctx, rotated); serr != nil {
			return "", fmt.Errorf("realmid token manager: persist rotated refresh token: %w", serr)
		}
	}

	exp := m.now().Add(time.Duration(mr.ExpiresIn) * time.Second)
	if mr.ExpiresIn == 0 {
		exp = m.now().Add(5 * time.Minute) // SPEC §4.0 default
	}
	m.mu.Lock()
	m.accessToken = mr.AccessToken
	m.accessExpiresAt = exp
	m.mu.Unlock()
	return mr.AccessToken, nil
}

// HandleStale answers an ADR-107 `token_stale` 401 by forcing ONE refresh and
// returning the token to replay the original request with.
//
// D13, the loop-breaker. D8 makes the C5 loop very unlikely by stamping the
// marker early; this makes it impossible, and it lives in the client SDK
// because that is where the retry decision is actually made. A second
// token_stale on a token this method already produced is a HARD failure —
// returned as token_stale so the caller can surface it, never retried.
//
// staleToken is the token the failing request carried. Passing a token that is
// no longer the manager's current one is not an error: another goroutine
// already refreshed, and the caller simply gets the newer token to replay with,
// costing the issuer nothing.
//
// Usage:
//
//	tok, err := tm.AccessToken(ctx)
//	res, err := call(tok)
//	if realmid.IsTokenStale(err) {
//	    tok, err = tm.HandleStale(ctx, tok)   // hard-fails on the second try
//	    if err != nil { return err }          // surface it; do NOT loop
//	    res, err = call(tok)
//	}
func (m *TokenManager) HandleStale(ctx ctxpkg.Context, staleToken string) (string, error) {
	m.mu.Lock()
	if staleToken != "" && staleToken == m.forcedToken {
		m.mu.Unlock()
		return "", &RealmError{
			Code:       ErrCodeTokenStale,
			HTTPStatus: 401,
			Message: "realmid: token_stale on a token minted by the previous forced refresh — " +
				"refusing to refresh again (ADR-107 D13). The authority marker is ahead of the " +
				"issuer's clock, or the subject is being re-marked faster than a token can be used.",
		}
	}
	// Someone else already refreshed past the failing token — replay with what
	// we hold rather than spending a mint.
	if staleToken != "" && m.accessToken != "" && staleToken != m.accessToken {
		tok := m.accessToken
		m.mu.Unlock()
		return tok, nil
	}
	// Drop the cached token so AccessToken cannot hand back the stale one.
	m.accessToken = ""
	m.accessExpiresAt = time.Time{}
	m.mu.Unlock()

	tok, err := m.AccessToken(ctx)
	if err != nil {
		return "", err
	}
	m.mu.Lock()
	m.forcedToken = tok
	m.mu.Unlock()
	return tok, nil
}
